src/base/static/state/ducks/map.js

```diff
@@ -1,311 +1,581 @@
+    .filter(lg => !!lg.basemap)
+    .map(lg => lg.mapboxLayers.concat(lg.mapboxFocusedLayers || []))
+    .reduce((flat, toFlatten) => flat.concat(toFlatten), [])
+    .map(layer => layer.id);
+
+  const interactiveLayerIds = mapConfig.layerGroups
+    .filter(lg => !!lg.interactive)
+    .map(lg => lg.mapboxLayers.concat(lg.mapboxFocusedLayers || []))
+    .reduce((flat, toFlatten) => flat.concat(toFlatten), [])
+    .map(layer => layer.id);
+
+  return {
+    type: LOAD_STYLE_AND_METADATA,
+    payload: {
+      style,
+      sourcesMetadata,
+      layerGroupsMetadata,
+      basemapLayerIds,
+      interactiveLayerIds,
+    },
+  };
+}
+
+export const mapViewportPropType = PropTypes.shape({
+  height: PropTypes.number.isRequired,
+  width: PropTypes.number.isRequired,
+  latitude: PropTypes.number,
+  longitude: PropTypes.number,
+  zoom: PropTypes.number,
+  pitch: PropTypes.number,
+  bearing: PropTypes.number,
+  transitionInterpolator: PropTypes.object,
+  transitionDuration: PropTypes.number,
+  transitionEasing: PropTypes.func,
+  minZoom: PropTypes.number.isRequired,
+  maxZoom: PropTypes.number.isRequired,
+});
+
+export const mapStylePropType = PropTypes.shape({
+  version: PropTypes.number,
+  name: PropTypes.string,
+  sources: PropTypes.object,
+  layers: PropTypes.arrayOf(PropTypes.object),
+});
+
+// Reducers:
+// TODO(luke): refactor our current implementation in AppView to use
 const INITIAL_STATE = {
-  visibleBasemapId: undefined,
-  layerStatuses: {},
-  featureFilters: [],
-  updatingFilterGroupId: undefined,
-  updatingFilterTargetLayer: undefined,
-  mapPosition: undefined,
-  isMapSizeValid: true,
-  isMapDragging: false,
+  viewport: {
+    width: 0,
+    height: 0,
+    minZoom: 1,
+    maxZoom: 18,
+    transitionInterpolator: new FlyToInterpolator(),
+    transitionEasing: easeInOutCubic,
+  },
+  style: {
+    version: 8,
+    name: "Mapseed",
+    glyphs: "mapbox://fonts/mapbox/{fontstack}/{range}.pbf",
+    sprite: `${window.location.protocol}//${
+      window.location.host
+    }/static/css/images/markers/spritesheet`,
+    sources: {},
+    layers: [],
+  },
+  layerGroupsMetadata: {},
+  sourcesMetadata: {},
+  basemapLayerIds: [],
+  interactiveLayerIds: [],
+  isMapSizeValid: false,
   isMapDragged: false,
+  isMapDragging: false,
 };
 
 export default function reducer(state = INITIAL_STATE, action) {
-  let newStatuses;
-
   switch (action.type) {
-    case UPDATE_MAP_DRAGGING:
+    case UPDATE_GEOJSON_FEATURES:
       return {
         ...state,
-        isMapDragging: action.payload,
-      };
-    case SET_MAP_SIZE_VALIDITY:
-      return {
-        ...state,
-        isMapSizeValid: action.payload,
+        style: {
+          ...state.style,
+          sources: {
+            ...state.style.sources,
+            [action.payload.sourceId]: {
+              ...state.style.sources[action.payload.sourceId],
+              data: {
+                ...state.style.sources[action.payload.sourceId].data,
```

Comment:
Is there a better pattern for updating deeply nested properties using the spread operator? This looks wrong to me and I also wonder what the performance implications are.

Reply:
Sounds good. I'm not sure we should mess too much with the structure of `sources`, though, since it's part of the style object that's plugged into the `react-map-gl`. But if the nested spread operators are just a cosmetic issue than I think it's fine.

Reply:
>I'm not sure we should mess too much with the structure of `sources`, though, since it's part of the style object that's plugged into the `react-map-gl`

we shouldn't let the Mapbox `style` object dictate our app's data schema, especially if preserving it is causing us pain. It seems to be the cause of the complexity described in this comment:
https://github.com/mapseed/platform/pull/1173#discussion_r258331098
where we are duplicating information between the `style` object, and the "metadata" object, and having to keep them in sync.

So it could be reasonable to structure the data in a way that best suites our app, and re-assembling it into a mapbox style object via selector. Type-checking via proptypes would ensure we are re-assembling the mapbox style object properly.

That's why I proposed the `combineReducers` solution in that thread above. And we don't have to make any changes now, unless you think the code is too complex or brittle. I just wanted to suggest that there might be a better way.

Reply:
I think that's a good perspective to take-- not letting an external library's requirements dictate our app's data structures.

The only thing that gives me pause is that the `style` state stored in Redux is accessed extremely frequently, so I wonder what a transformation on every selection would do to performance. But that might be premature optimization...

Reply:
I think the perf issue can be curbed by using reselect: https://github.com/reduxjs/reselect :thinking: 

It might be worth filing an issue, or adding a comment to clarify why we have to keep our `style` and `metadata` data structures in sync.

Either way, this PR is an overall great improvement, so I'm happy with punting these changes for now :+1: